File: sc-processor/src/main/java/com/e16din/sc/processor/code/OnHideViewControllerGenerator.java
package com.e16din.sc.processor.code;

import com.e16din.sc.annotations.OnHide;
import com.squareup.javapoet.MethodSpec;

import java.util.HashMap;
import java.util.Map;
import java.util.Set;

import javax.annotation.processing.RoundEnvironment;
import javax.lang.model.element.Element;
import javax.lang.model.element.ElementKind;
import javax.lang.model.element.ExecutableElement;
import javax.lang.model.element.Modifier;

public class OnHideViewControllerGenerator {

    private static Map<String, String> methodsMap = new HashMap<>();// <controller name , result method>


    public static MethodSpec process(RoundEnvironment roundEnvironment) {
        Set<? extends Element> elementsAnnotatedWithOnMenuItemClick =
                roundEnvironment.getElementsAnnotatedWith(OnHide.class);
        for (Element element : elementsAnnotatedWithOnMenuItemClick) {
            if (element.getKind() != ElementKind.METHOD) {
                return null;
            }

            ExecutableElement execElement = (ExecutableElement) element;

            final String viewControllerName = execElement.getEnclosingElement().toString();
            final String methodName = execElement.getSimpleName().toString();

            methodsMap.put(viewControllerName, methodName);
        }

        return MethodSpec
                .methodBuilder("onHideViewController")
                .addModifiers(Modifier.PUBLIC)
                .addParameter(Object.class, "vc")
                .addCode(createCode())
                .build();

    }

    private static String createCode() {
        Set<String> viewControllers = methodsMap.keySet();

        StringBuilder code = new StringBuilder();

        code.append("    final String vcName = com.e16din.sc.Utils.getClassDefaultName(vc);\n");
        code.append("    switch (vcName) {\n");


        for (String vcName : viewControllers) {
            code.append("            case \"")
                    .append(vcName)
                    .append("\":\n")
                    .append("((")
                    .append(vcName)
                    .append(") vc).")
                    .append(methodsMap.get(vcName))
                    .append("();\n")
                    .append("break;\n\n");
        }

        code.append("}\n");


        return code.toString();
    }
}

File: sc-processor/src/main/java/com/e16din/sc/processor/code/OnHideViewControllerGenerator.kt
package com.e16din.sc.processor.code

import com.e16din.sc.annotations.OnHide
import com.e16din.sc.processor.ViewControllerProcessor
import com.squareup.kotlinpoet.FunSpec
import com.squareup.kotlinpoet.KModifier
import java.util.*
import javax.annotation.processing.RoundEnvironment
import javax.lang.model.element.ElementKind
import javax.lang.model.element.ExecutableElement

object OnHideViewControllerGenerator {

    private val methodsMap = HashMap<String, String>()// <controller name , result method>


    fun process(roundEnvironment: RoundEnvironment): FunSpec? {
        val elementsAnnotatedWithOnMenuItemClick = roundEnvironment.getElementsAnnotatedWith(OnHide::class.java)
        for (element in elementsAnnotatedWithOnMenuItemClick) {
            if (element.kind != ElementKind.METHOD) {
                return null
            }

            val execElement = element as ExecutableElement

            val viewControllerName = execElement.enclosingElement.toString()
            val methodName = execElement.simpleName.toString()

            methodsMap[viewControllerName] = methodName
        }

        return FunSpec
                .builder("onHideViewController")
                .addModifiers(KModifier.OVERRIDE, KModifier.PUBLIC)
                .addParameter("vc", Any::class)
                .addCode(createCode())
                .build()

    }

    private fun createCode(): String {
        val viewControllers = methodsMap.keys

        val code = StringBuilder()

        code.append("    var vcName = " + ViewControllerProcessor.SC_PACKAGE + ".Utils.getClassDefaultName(vc);\n")
        code.append("    when (vcName) {\n")

        for (vcName in viewControllers) {
            code.append("            \"")
                    .append(vcName)
                    .append("\"->\n")
                    .append("(vc as $vcName).")
                    .append(methodsMap[vcName])
                    .append("();\n")
        }

        code.append("}\n")

        return code.toString()
    }
}
